session/schema_amender.go

```diff
@@ -331,38 +323,77 @@ func (a *amendOperationAddIndexInfo) String() string {
 		}
-		if !isDeleteOp(keyOp) {
-			continue
+		newIdxMutation.Reset()
+		oldIdxMutation.Reset()
+		keyOp := commitMutations.GetOps()[i]
+		if addIndexNeedRemoveOp(a.info.AmendOpType) {
+			if isDeleteOp(keyOp) {
+				err := a.genOldIdxKey(ctx, sctx, key, kvMap.oldRowKvMap, oldIdxMutation)
+				if err != nil {
+					return err
+				}
+			}
 		}
-		err := a.processRowKey(ctx, sctx, key, kvMap.oldRowKvMap, resAddMutations)
-		if err != nil {
-			return err
+		if addIndexNeedAddOp(a.info.AmendOpType) {
+			if isInsertOp(keyOp) {
+				err := a.genNewIdxKey(ctx, sctx, key, kvMap.newRowKvMap, newIdxMutation)
+				if err != nil {
+					return err
+				}
+			}
 		}
-	}
-	return nil
-}
-
-func (a *amendOperationAddNewIndex) genMutations(ctx context.Context, sctx sessionctx.Context, commitMutations tikv.CommitterMutations,
-	kvMap *rowKvMap, resAddMutations *tikv.CommitterMutations) error {
-	for i, key := range commitMutations.GetKeys() {
-		keyOp := commitMutations.GetOps()[i]
-		if tablecodec.IsIndexKey(key) || tablecodec.DecodeTableID(key) != a.info.tblInfoAtCommit.Meta().ID {
-			continue
+		skipMerge := false
+		if a.info.AmendOpType == AmendNeedAddDeleteAndInsert {
+			// If the old index key is the same with new index key, then the index related row value
+			// is not changed in this row, we don't need to add or remove index keys for this row.
+			if len(oldIdxMutation.GetKeys()) > 0 && len(oldIdxMutation.GetKeys()) == len(newIdxMutation.GetKeys()) {
+				if bytes.Equal(oldIdxMutation.GetKeys()[0], newIdxMutation.GetKeys()[0]) {
+					skipMerge = true
+				}
+			}
 		}
-		if !isInsertOp(keyOp) {
-			continue
+		if !skipMerge {
+			if len(oldIdxMutation.GetKeys()) > 0 {
+				deletedMutations.MergeMutations(*oldIdxMutation)
+			}
+			if len(newIdxMutation.GetKeys()) > 0 {
+				insertedMutations.MergeMutations(*newIdxMutation)
+			}
 		}
-		err := a.processRowKey(ctx, sctx, key, kvMap.newRowKvMap, resAddMutations)
-		if err != nil {
-			return err
+	}
+	// For unique index, there may be conflicts on the same unique index key from different rows.Consider a update statement,
+	// "Op_Del" on row_key = 3, row_val = 4, the "Op_Del"     unique_key_4 -> nil will be generated.
+	// "Op_Put" on row_key = 0, row_val = 4, the "Op_Insert"  unique_key_4 -> 0   will be generated.
+	// The "Op_Insert" should cover the "Op_Del" otherwise the new put row value will not have a correspond index value.
+	if a.info.indexInfoAtCommit.Meta().Unique {
+		for i := 0; i < len(deletedMutations.GetKeys()); i++ {
+			key := deletedMutations.GetKeys()[i]
+			if _, ok := a.insertedNewIndexKeys[string(key)]; !ok {
+				resAddMutations.Push(deletedMutations.GetOps()[i], key, deletedMutations.GetValues()[i], deletedMutations.GetPessimisticFlags()[i])
+			}
+		}
+		for i := 0; i < len(insertedMutations.GetKeys()); i++ {
+			key := insertedMutations.GetKeys()[i]
+			destKeyOp := pb.Op_Insert
+			if _, ok := a.deletedOldIndexKeys[string(key)]; ok {
+				destKeyOp = pb.Op_Put
```

Comment:
If there is an operation performs `Op_Del` after `Op_Insert`, will it be changed into `Op_Put`?

Reply:
Yes, the related scenerio is in the above [comment](https://github.com/pingcap/tidb/pull/20904#issuecomment-724057626), in this case, 
```
Op_Del old_row_key = 3,                                      -> generate Op_Del unique index key 4 -> nil
Op_Put new_row_key = 5 row_value = 4              -> generate Op_Insert on unique index key 4 -> 5
 ```